controllers/iochaos_controller.go

```diff
@@ -14,37 +14,62 @@
 package controllers
 
 import (
-	"github.com/go-logr/logr"
+	"context"
 
-	chaosmeshv1alpha1 "github.com/pingcap/chaos-mesh/api/v1alpha1"
+	"github.com/pingcap/chaos-mesh/api/v1alpha1"
 	"github.com/pingcap/chaos-mesh/controllers/iochaos"
+	"github.com/pingcap/chaos-mesh/pkg/utils"
 
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // IoChaosReconciler reconciles a IoChaos object
 type IoChaosReconciler struct {
 	client.Client
-	Log logr.Logger
+	Log      logr.Logger
+	Recorder record.EventRecorder
 }
 
 // +kubebuilder:rbac:groups=chaosmesh.pingcap.com,resources=iochaos,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=chaosmesh.pingcap.com,resources=iochaos/status,verbs=get;update;patch
 
-func (r *IoChaosReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+// Reconcile reconciles an IOChaos resource
+func (r *IoChaosReconciler) Reconcile(req ctrl.Request) (result ctrl.Result, err error) {
 	logger := r.Log.WithValues("iochaos", req.NamespacedName)
 
 	reconciler := iochaos.Reconciler{
 		Client: r.Client,
 		Log:    logger,
 	}
+	chaos := &v1alpha1.IoChaos{}
+	if err := r.Get(context.Background(), req.NamespacedName, chaos); err != nil {
+		r.Log.Error(err, "unable to get iochaos")
+		return ctrl.Result{}, nil
+	}
 
-	return reconciler.Reconcile(req)
+	result, err = reconciler.Reconcile(req, chaos)
+	if !chaos.IsDeleted() {
+		if err != nil {
+			r.Recorder.Event(chaos, v1.EventTypeWarning, utils.EventChaosInjectFailed, err.Error())
+		} else {
+			r.Recorder.Event(chaos, v1.EventTypeNormal, utils.EventChaosStarted, "")
+		}
+	} else {
+		if err != nil {
+			r.Recorder.Event(chaos, v1.EventTypeWarning, utils.EventChaosRecoverFailed, err.Error())
+		} else {
+			r.Recorder.Event(chaos, v1.EventTypeNormal, utils.EventChaosCompleted, "")
+		}
+	}
+	return result, nil
```

Comment:
For example, If a schedule Io chaos was created and start to be reconciled. The scheduler is every 2 min and the duration is 30s.
Without considering inject failing, I think the event should be recorded like：
```bash
0:00 iochaos xxx start performing
0:30 iochaos xxx recovered
2:00 iochaos xxx start performing
2:30 iochaos xxx recovered
```
instaed of recording `EventChaosStarted` every resync duration.

Reply:
Maybe add more detailed message for `EventChaosStarted`?

Reply:
> Maybe add detailed message for `EventChaosStarted`?

But it would still recorded each resync duraiton (default every 10 seconds) with lots of useless requeue infomartion.

Reply:
Yes. For each round of chaos, there will be a starting event. I think that is OK. How do you think? @zhouqiang-cl

Reply:
It is ok for me. Each round of chaos, we just need 1 record

Reply:
OK. I think we could just leave it that way for now. With the system grown, it maybe get  clearer for how to records more friendly events. Are we agreed with this ? @Yisaer

Reply:
@AngleNet I think @zhouqiang-cl misunderstand it. I pull your branch and test it which confirmed my guess. The controlled would record event nearly every 10 sec as I showed below. As Chaos User, I think the chaos action perform and recovery information would more useful with my own monitoring system during chaos engineering test.

![x](https://user-images.githubusercontent.com/13427348/74827314-2ca54200-5348-11ea-989a-bb14b67a6be2.png)

Reply:
OK. I will check it later.

Reply:
I am full of expectation ❤️

Reply:
@Yisaer You are right about it :+1: :+1:  . I have already checked that. For chaos with single round, I think it's fine. However, the events acts just like you figured. I'm trying to make it right.

Reply:
I find that K8s sometimes groups same events. It will break the timeline of the events. Is that ok ? @Yisaer @zhouqiang-cl

Reply:
You mean `(x11 over 3m4s)`, that is ok

Reply:
Yes. All right.